blockchain/types/transaction.go

```diff
@@ -254,8 +256,44 @@ func (tx *Transaction) UnmarshalJSON(input []byte) error {
 
 func (tx *Transaction) Gas() uint64        { return tx.data.GetGasLimit() }
 func (tx *Transaction) GasPrice() *big.Int { return new(big.Int).Set(tx.data.GetPrice()) }
-func (tx *Transaction) Value() *big.Int    { return new(big.Int).Set(tx.data.GetAmount()) }
-func (tx *Transaction) Nonce() uint64      { return tx.data.GetAccountNonce() }
+func (tx *Transaction) GasTipCap() *big.Int {
+	if tx.Type() == TxTypeDynamicFee {
+		te := tx.GetTxInternalData().(TxInternalDataEthTyped)
+		return te.GetGasTipCap()
+	}
+
+	return tx.data.GetPrice()
+}
+
+func (tx *Transaction) GasFeeCap() *big.Int {
+	if tx.Type() == TxTypeDynamicFee {
+		te := tx.GetTxInternalData().(TxInternalDataEthTyped)
+		return te.GetGasFeeCap()
+	}
+
+	return tx.data.GetPrice()
+}
+
+func (tx *Transaction) EffectiveGasTip(baseFee *big.Int) *big.Int {
+	if tx.Type() == TxTypeDynamicFee {
+		te := tx.GetTxInternalData().(TxInternalDataEthTyped)
+		return math.BigMin(te.GetGasTipCap(), new(big.Int).Sub(te.GetGasFeeCap(), baseFee))
+	}
+
+	return tx.GasPrice()
+}
+
+func (tx *Transaction) EffectiveGasPrice(baseFee *big.Int) *big.Int {
+	if tx.Type() == TxTypeDynamicFee {
+		te := tx.GetTxInternalData().(TxInternalDataEthTyped)
+		return math.BigMin(new(big.Int).Add(te.GetGasTipCap(), baseFee), te.GetGasFeeCap())
```

Comment:
Don't we need to check effectiveGasPrice is same with gasPrice somewhere?

Reply:
This function calculates `effectiveGasPrice` according to the concept. And when checking transaction validation in TxPool, gas related fields(max_priority_fee, max_fee) are being checked.